euphoria-flink/src/main/java/cz/seznam/euphoria/flink/streaming/ReduceStateByKeyTranslator.java

```diff
@@ -44,114 +47,124 @@
   public DataStream<?> translate(FlinkOperator<ReduceStateByKey> operator,
                                  StreamingExecutorContext context)
   {
-    DataStream<?> input =
+    DataStream input =
             Iterables.getOnlyElement(context.getInputStreams(operator));
 
     ReduceStateByKey origOperator = operator.getOriginalOperator();
 
     StateFactory<?, State> stateFactory = origOperator.getStateFactory();
+    CombinableReduceFunction stateCombiner = origOperator.getStateCombiner();
+
+    Windowing windowing = origOperator.getWindowing();
+    if (windowing == null) {
+      // use attached windowing when no windowing explicitly defined
+      windowing = new AttachedWindowing<>();
+    }
 
-    final Windowing windowing = origOperator.getWindowing();
     final UnaryFunction keyExtractor = origOperator.getKeyExtractor();
     final UnaryFunction valueExtractor = origOperator.getValueExtractor();
     final UnaryFunction eventTimeAssigner = origOperator.getEventTimeAssigner();
 
-    FlinkStreamingStateStorageProvider storageProvider
-        = new FlinkStreamingStateStorageProvider();
-
-    DataStream<StreamingWindowedElement<?, Pair>> folded;
-    // apply windowing first
-    if (windowing == null) {
-      WindowedStream windowed =
-          context.attachedWindowStream((DataStream) input, keyExtractor, valueExtractor);
-      // equivalent operation to "left fold"
-      folded = windowed.apply(new RSBKWindowFunction(storageProvider, stateFactory))
-          .name(operator.getName())
-          .setParallelism(operator.getParallelism());
-    } else {
-      WindowedStream<MultiWindowedElement<?, Pair>, Object, FlinkWindow>
-          windowed = context.flinkWindow(
-              (DataStream) input, keyExtractor, valueExtractor, windowing, eventTimeAssigner);
-      // equivalent operation to "left fold"
-      folded = windowed.apply(new RSBKWindowFunction(storageProvider, stateFactory))
-          .name(operator.getName())
-          .setParallelism(operator.getParallelism());
+    if (eventTimeAssigner != null) {
+      input = input.assignTimestampsAndWatermarks(
+              new EventTimeAssigner(context.getAllowedLateness(), eventTimeAssigner));
     }
 
+    // assign windows
+    DataStream<KeyedMultiWindowedElement> windowed = input.map(new WindowAssigner(
+            windowing, keyExtractor, valueExtractor, eventTimeAssigner))
+            .setParallelism(operator.getParallelism());
+
+    DataStream<WindowedElement<?, Pair>> reduced = (DataStream) windowed.keyBy(new KeyExtractor())
```

Comment:
i think we're better off with `.keyBy("key")`; flink will then derive the type information from the input data stream automatically; it'll still be "object" at this moment since we don't supply enough type information through the `WindowAssigner`, but that's about to come in some future.